lib/src/components/AbstractDialog/AbstractDialog.tsx

```diff
@@ -0,0 +1,213 @@
+ */
+import React, {
+  forwardRef,
+  KeyboardEvent,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState
+} from 'react';
+
+import { omit } from 'ramda';
+import { Props, AbstractDialogContextValue } from './types';
+
+export const AbstractDialogContext = React.createContext<AbstractDialogContextValue | null>(null);
+
+// Typescript has rather obnoxiously partially removed support for HTMLDialogElement.
+// See https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-907490128 and
+// https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-968299542 . This even though
+// Firefox appears to intend to finally add support in v98. Until TS restores this, we'll have to jump
+// through some hoops...
+const hasWindow = typeof window !== 'undefined',
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    dynamicallyTypedWindow = hasWindow && window as any,
+    hasNativeModalSupport = !!(hasWindow && dynamicallyTypedWindow.HTMLDialogElement &&
+      dynamicallyTypedWindow.HTMLDialogElement.prototype.showModal);
+
+const createCancelEvent = () => new Event('cancel', { cancelable: true });
+
+/**
+ * Abstracted Dialog element implementation and behaviors.
+ * @param className - A classname string for the dialog element.
+ * @param onCancel - A callback function that gets called when the dialog is canceled.
+ * @param cancelOnClickOutside - A boolean that if true triggers onCancel when the user clicks outside the dialog.
+ * @param cancelOnClickOutsideTargetClassName - A classname, if specified, selects an element inside the dialog
+ *    element that becomes the reference element for when the user clicks outside of, triggers the cancelOnClickOutside.
+ * @param useNativeCancelOnEscape - If this is set to true, it will attempt to use native dialog element
+ *    cancel behavior when escape is pressed. By default this is set to false, which means it uses event.preventDefault
+ *    and calls onCancel callback when escape is pressed.
+ * @return - Abstracted dialog element.
+ */
+
+// propTypes static analysis doesn't work with the way this component is written
+/* eslint-disable react/prop-types */
+const AbstractDialog = forwardRef<HTMLDialogElement, Props>((props, ref) => {
+  const {
+    className,
+    children,
+    cancelOnClickOutside,
+    onCancel,
+    useNativeCancelOnEscape,
+    role,
+    cancelOnClickOutsideTargetClassName,
+    ...attrs
+  } = props;
+
+  const dialogRef = useRef<HTMLDialogElement>(null);
+
+  // The dialogRef value needs to get passed down in a context. But the context needs to know when the ref
+  // value has updated, and refs aren't tracked like state values. So we have to copy the ref value into a state
+  // value in order for it to be tracked.
+  const [dialogRefState, setDialogRefState] = useState<HTMLDialogElement | null>(null);
+
+  useImperativeHandle(ref, () => dialogRef.current as HTMLDialogElement);
```

Comment:
What is this line doing?

Reply:
> What is this line doing?

Is it just copying the ref value so we can use it interally as `dialogRef` and externally as `ref`? We usually use `useMergedRef` for that

Reply:
> > What is this line doing?
> 
> Is it just copying the ref value so we can use it interally as `dialogRef` and externally as `ref`? We usually use `useMergedRef` for that

Fixed here: 1ff7def85a3aa1a4ce3f9a33fc6a0d9927eca84b

Reply:
> Fixed here: https://github.com/sonatype/sonatype-react-shared-components/commit/1ff7def85a3aa1a4ce3f9a33fc6a0d9927eca84b

That can't be working.  The usage of `useMergedRef` is to create and return a new ref which, when set, sets all of the refs that you passed in to `useMergedRef`.  Calling `useMergedRef` without capturing its return value is effectively a noop.

Reply:
> > Fixed here: [1ff7def](https://github.com/sonatype/sonatype-react-shared-components/commit/1ff7def85a3aa1a4ce3f9a33fc6a0d9927eca84b)
> 
> That can't be working. The usage of `useMergedRef` is to create and return a new ref which, when set, sets all of the refs that you passed in to `useMergedRef`. Calling `useMergedRef` without capturing its return value is effectively a noop.

Oh we fixed that in a subsequent commit but forgot to update the comment:
Here's the commit: fab5fe1ae780213af9632e1dad016203e89491b5